homeassistant/components/switch/w800rf32.py

```diff
@@ -0,0 +1,142 @@
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ['w800rf32']
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_DEVICES, default={}): {
+        cv.string: vol.Schema({
+            vol.Optional(CONF_NAME): cv.string,
+            vol.Optional(CONF_FIRE_EVENT, default=False): cv.boolean
+        })
+    },
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the Switch platform to w800rf32."""
+    import W800rf32 as w800rf32mod
+    switches = []
+
+    # device_id --> "c1 or a3" X10 device. entity (type dictionary) --> name, device_class etc
+    for device_id, entity in config[CONF_DEVICES].items():
+
+        if device_id in w800rf32.W800_DEVICES:
+            continue
+
+        _LOGGER.debug("Add %s w800rf32.switch",
+                      entity[ATTR_NAME])
+
+        event = None  # None until an event happens
+        device = W800rf32Switch(event, entity.get(CONF_NAME), entity[CONF_FIRE_EVENT],)
+
+        switches.append(device)
+        w800rf32.W800_DEVICES[device_id] = device
+
+    add_entities(switches)
+
+    def switch_update(event):
+        """Call for control updates from the w800rf32 gateway."""
+        switch = None
+
+        if not isinstance(event, w800rf32mod.W800rf32Event):
+            return
+
+        # make sure it's lowercase
+        device_id = event.device.lower()
+
+        # get the name, ex: motion_hall
+        if device_id in w800rf32.W800_DEVICES:
+            switch = w800rf32.W800_DEVICES[device_id]
+
+        if switch is None:
+            return
+
+        if not isinstance(switch, W800rf32Switch):
+            return
+        else:
+            _LOGGER.debug(
+                "Binary sensor update (Device ID: %s Class: %s)",
+                event.device,
+                event.device.__class__.__name__)
+
+        w800rf32.apply_received_command(event)
+
+    # Subscribe to main w800rf32 events
+    if switch_update not in w800rf32.RECEIVED_EVT_SUBSCRIBERS:
+        w800rf32.RECEIVED_EVT_SUBSCRIBERS.append(switch_update)
+
+
+class W800rf32Switch(SwitchDevice):
+    """A representation of a w800rf32 switch."""
+
+    def __init__(self, event, name, should_fire=False,):
+        """Initialize the w800rf32 switch."""
+        self.event = event
+        self._name = name
+        self._should_fire_event = should_fire
+        self._state = False
+
+    @property
+    def name(self):
+        """Return the device name."""
+        return self._name
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def is_on(self):
+        """Return true if the sensor state is True."""
+        return self._state
+
+    @property
+    def should_fire_event(self):
+        """Return is the device must fire event."""
+        return self._should_fire_event
+
+    def turn_on(self, **kwargs):
+        self._state = True
+        self.schedule_update_ha_state()
```

Comment:
If the state update callback is fast enough, don't update state in the actuator method, but let the callback do that.

Reply:
Unfortunately if this is not used, when you toggle the web interface on, it will immediately turn it's off.
With this in place everything works as it should.

Reply:
It should not immediately turn off. There should be a 2 sec wait. If it does immediately turn off, then something else is weird.

Reply:
It should never turn off.  It should only turn off when you go back and manually toggle it.  It's a switch, you turn it on, it should stay on, you turn it off, it should stay off.  I was following the example of the demo.py switch and also mqtt.  should I instead just call my own update?

Reply:
Yes, it should stay on. The 2 seconds is a timeout that starts when pressing the switch in the frontend. If the state hasn't been updated after the two seconds, the switch will move back to the previous position, indicating that the service failed to change state of the switch.

So if the state update callback comes in within 2 seconds, that should be enough to update the state and have the switch stay in the new position.

Does this make sense?

Reply:
Well it makes sense but what is the flow of information between the frontend and the switch device?
I'm not clear on that.  From what I see it calls the <device>.turn_on() or turn_off() command, which from what you are saying should just set the <device>._state property.   What is the frontend looking for to ensure the state stays toggled?

Reply:
The frontend is waiting for a state update in the state machine. This is what we schedule with `schedule_update_ha_state`. But instead of doing that within the actuator method, it's better to let the device callback and schedule a state update. That way we know that the device has actually changed state. Of course, if this is feasible depends on the device, and how responsive it is.

So that's why I asked how fast the callback from the device is called after the service call.

Reply:
Well possibly it's not fast enough, I'm not sure at this point.  It does all work well by calling the schedule_update_ha_state from the turn_on() and turn_off() methods.  I think I'll shelve this minor issue until I sort out the EVENT_SUBSCRIBER code you listed.

Reply:
Sounds good. Although I wouldn't call it a minor issue. It's kind of fundamental to change the state on the device when calling the service of the switch. If we can get feedback of that, the integration is a lot more useful.

Reply:
I notice there are quite a few SWITCH platforms that do this.  broadlink, command_line, demo, mqtt, mysensors, pilight,raspihats, rpi_gpio and rpi_pfio to name a few.

Reply:
That is exactly the case with this device, there is no feed back.  It's a one way rf device.